Minor changes

BrianBohe opened the pull request:
Adding special tokens, two new extensions that are on JIT/Directed, and an index strategy for class deduplication when previous deduplicating strategies fail.

BrianBohe committed:
Adding more special tokens

BrianBohe committed:
Adding two more test extensions

BrianBohe committed:
Extending classname deduplicating strategy

Using indexes when previous strategies fail.

BrianBohe committed:
Extending classname replace on use as event

The syntax: " .event aClass aVariableName" should match any
replacement of "aClass" with "aNewNamespace.aClass".

BrianBohe committed:
Extending class def detection

The line ".class private serializable sealed Days" is valid syntax
but Days would not be considered as a class before this commit.

BrianBohe committed:
Fixing typo on regex

BrianBohe committed:
Including last C# using line on check

BrianBohe committed:
Extending namespace replacement search space

markples commented on ILTransform/ILRewriter.cs:
```diff
@@ -378,7 +378,8 @@ private void RewriteFile(string source)
                             string qualifiedClassName = _testProject.DeduplicatedNamespaceName + "." + className!;
-                            for (int s = lineIndex; s < lines.Count; s++)
+                            // There could be a class usage before the namespace line
+                            for (int s = 0; s < lines.Count; s++)
```
The logic for finding `NamespaceLine` is a bit odd.. it looks like it is trying to find the beginning of the method/type that the entry point is in, but then we extend it to the end of the file.  It's not probably necessary to do anything about it, but it's just inconsistent to include extra things after the entry point but not before it.

Since `NamespaceLine` can be in the middle of the file, this change definitely makes sense.

I don't believe the `s != i` is needed anymore.  `ReplaceIdent(..., IdentKind.TypeUse)` should be fine to avoid renaming the class definition itself.  (and technically line-based is probably not the correct way anyway, though it seems pretty reasonable to believe that our tests have newlines after the .class start.

markples commented on ILTransform/ILRewriter.cs:
```diff
@@ -417,8 +418,8 @@ private void RewriteFile(string source)
-            Regex usingXunitRegex = new Regex(@"^\s*using\s+Xunit\s*(?:\\,*)?;");
-            bool usingXUnit = (_testProject.LastUsingLine >= 0 && lines.Take(_testProject.LastUsingLine).Any(l => usingXunitRegex.IsMatch(l)));
+            Regex usingXunitRegex = new Regex(@"^\s*using\s+Xunit\s*(?:\\.*)?;");
+            bool usingXUnit = (_testProject.LastUsingLine >= 0 && lines.Take(_testProject.LastUsingLine+1).Any(l => usingXunitRegex.IsMatch(l)));
```
`LastUsingLine` is based on `line.StartsWith("using")` - not a very robust check and one of the things that this regex does better.  The potential problem being solved by the `Take` is that `using xx` could appear in the file (for example in a string literal).

I don't understand the `+1` here.  Do you have a case where there are "normal" usings but then the xunit one was missed when setting `LastUsingLine`?  Or is the line number off by one due to an insertion?  I'm asking here because `+1` seems like it might be covering a larger problem.

(Of course, the regex is more robust but still not an IL parser, which would be necessary to get this fully right, so we are approximating with whatever we do.)

BrianBohe replied:
I understand most (or all) indexes are 1-indexed while the array of strings which contains the test file lines is 0-indexed. So the +1 is to contain the last using.

BrianBohe replied:
It was adding "using Xunit;" when it was already in the file on jit/directed

markples replied:
I think everything in ILTransform is 0-based.  It's all using that array (well, `List`) of strings.

So I think something unexplained was going on in your example.  AddILFactAttributes can modify the entrypoint before getting to this code, but I don't think that code is going to be before `using` where line number changes would matter.  Could `LastUsingLine` be incorrect?  It is simply using `line.StartsWith("using"), so whitespace or a comment at the beginning of the line would hide it.

BrianBohe replied:
File lines are 0-indexed in the code. Let's say "using Xunit;" is on line 3, then Take(3) would inspect only lines 0, 1 and 2.

markples commented on ILTransform/TestProject.cs:
```diff
@@ -511,6 +512,14 @@ public static void GetKeyNameRootNameAndSuffix(string path, out string keyName,
             {
                 suffixIndex -= 2;
             }
+            else if (rootName.EndsWith("_TargetUnix"))
+            {
+                suffixIndex -= 11;
+            }
+            else if (rootName.EndsWith("_TargetWindows"))
+            {
+                suffixIndex -= 14;
+            }
```
This brings up something that I don't know about our test system.  Do we build expect a particular build of src\test to be valid for all targets?  My guess would be that a particular build only supports one target - for example, there are lots of tests with C++ code that must be compiled to a specific architecture.

But that raises the question - why does this test have two project files instead of just one with an ifdef?

BrianBohe replied:
In this case there was a constant definition that would be defined for Windows but not for Unix systems, that is the reason why there are two different csproj files, but you are right, we can just make that Conditional on the target and have only one csproj.

BrianBohe replied:
I am reverting this commit change

markples commented on ILTransform/TestProject.cs:
```diff
@@ -2202,16 +2211,50 @@ string InterestingNamespace(TestProject project)
+                        else
+                        {
+                            string indexedName = string.IsNullOrEmpty(filename) ? projectName : filename;
+                            if (!filenameCounter.ContainsKey(indexedName))
+                            {
+                                filenameCounter[indexedName] = 1;
+                            }
+                            while (!dedupNames.Add(indexedName + filenameCounter[indexedName]++) && filenameCounter[indexedName] < 1500) { }
+                            bestAttempt.Add(indexedName + (filenameCounter[indexedName] - 1));
```
Why would the next index be in the names already?  Is this for handling something like (A\foo.ilproj, B\foo.ilproj, C\foo.ilproj, D\foo2.ilproj)?  What is special about 1500?  Do you need an error check if it gets to 1500, or are you replying on it eventually failing the uniqueness check below?

I think this could be simplified - something like
```
                            int counter = filenameCounter.GetValueOrDefault(indexedName, 1);
                            while (!dedupNames.Add(indexedName + counter) && filenameCounter[indexedName] < 1500) { counter++; }
                            filenameCounter[indexedName] = counter;
                            bestAttempt.Add(indexedName + counter);

BrianBohe replied:
It is just in case there are indexes already on proj files. The 1500 shouldn't be reached, we can remove it, I added it as a way of ensuring termination as well as getting notified (finding duplicated cases) if we reach that point.

BrianBohe committed:
Printing more information on -public

Printing file paths with [Fact] attribute that are not matching
patterns for main methods.

BrianBohe committed:
Revert "Adding two more test extensions"

This reverts commit 4f9a14a3befdf8b8bb560854a8dd067912bee185.

BrianBohe committed:
Refactoring deduplicating algorithm

BrianBohe committed:
Removing unnecessary check

BrianBohe committed:
Refactoring dedup algorithm
